File: JobDescription/src/src/repository/BaseRepository.java
package repository;

import domain.Entity;
import validator.Validator;


import java.util.*;

/**
 * Created by claudiu on 11.10.2016.
 */
public class BaseRepository<E extends Entity> implements Repository<E> {

    protected List<E> items;
    protected Validator<E> validator;

    public BaseRepository(Validator<E> validator) {
        this.validator = validator;
        this.items = new ArrayList<>();
    }

    public BaseRepository() {
    }

    public void add(E item) throws Exception {
        this.validator.validate(item);
        this.items.add(item);
    }

    public void remove(E item) throws Exception {
        this.validator.validate(item);
        this.items.remove(item);
    }

    public void update(E item) throws Exception {
        this.validator.validate(item);
        for (Entity entity : this.getAll()) {
            if (entity.getId() == item.getId()) {
                int index = this.items.indexOf(entity);
                this.items.set(index, item);
                return;
            }
        }
    }

    public List<E> getAll() {
        return this.items;
    }

    @SuppressWarnings("unchecked")
    public boolean existId(int id) {
        for (Entity entity : this.getAll()) {
            if (entity.getId() == id)
                return true;
        }
        return false;
    }

    public int getLastId() {
        if (this.items.size() == 0)
            return 0;
        return this.items.get(this.items.size() - 1).getId();
    }

    public Entity getEntityById(int id) {
        for (Entity entity : this.getAll()) {
            if (entity.getId() == id) {
                return entity;
            }
        }
        return null;
    }

    public void finalize() throws Throwable {
        super.finalize();
    }
}

File: JobDescriptionCsharp/JobDescriptionCsharp/repository/BaseRepository.cs
using JobDescriptionCsharp.domain;
using JobDescriptionCsharp.validator;
using System.Collections.Generic;
using System.Linq;

namespace JobDescriptionCsharp.repository
{
    public class BaseRepository<GenericType> : IRepository<GenericType> where GenericType : Entity
    {
        protected List<GenericType> items = new List<GenericType>();
        protected IValidator<GenericType> validator;

        public BaseRepository(IValidator<GenericType> validator)
        {
            this.validator = validator;
        }

        public void add(GenericType item)
        {
            validator.validate(item);
            items.Add(item);
        }

        public List<GenericType> getAll()
        {
            return items;
        }

        public GenericType getEntityById(int id)
        {
            var entity = from item in items
                         where item.id == id
                         select item;
            return entity.FirstOrDefault();
        }

        public int getLastId()
        {
            if (items.Count == 0)
                return 0;
            return items.ElementAt(items.Count - 1).id;
        }

        public void remove(GenericType item)
        {
            items.Remove(item);
        }

        public void update(GenericType item)
        {
            validator.validate(item);
            items.Insert(item.id - 1, item);
        }
    }
}
